core-processor/src/ext.rs

```diff
@@ -587,6 +591,48 @@ impl EnvExt for Ext {
         }
     }
 
+    fn reserve_gas(&mut self, amount: u32, blocks: u32) -> Result<ReservationId, Self::Error> {
+        self.charge_gas_runtime(RuntimeCosts::ReserveGas)?;
+
+        let common_charge = self.context.gas_counter.reduce(amount as u64);
+        let allowance_charge = self.context.gas_allowance_counter.charge(amount as u64);
+        self.check_charge_results(common_charge, allowance_charge)?;
+
+        let ProcessorContext {
+            message_context,
+            gas_reserver,
+            block_info,
+            ..
+        } = &mut self.context;
+
+        let msg_id = message_context.current().id();
+        let bn = block_info.height + blocks + 1;
+        let id = gas_reserver.reserve(msg_id, amount, bn);
+
+        Ok(id)
+    }
+
+    fn unreserve_gas(&mut self, id: ReservationId) -> Result<(), Self::Error> {
+        self.charge_gas_runtime(RuntimeCosts::UnreserveGas)?;
+
+        let amount = self
+            .context
+            .gas_reserver
+            .unreserve(id)
+            .ok_or_else::<Self::Error, _>(|| ExecutionError::InvalidReservationId.into())?;
+        let amount = amount as u64;
+
+        // this statement is like in `Self::refund_gas()` but it won't affect "burned" counter
+        // because we don't actually refund we just rise "left" counter during unreservation
+        if self.context.gas_counter.increase(amount) {
+            self.context.gas_allowance_counter.refund(amount);
```

Comment:
This possibly can lead to block production time overflow: if somebody will reserve too many gas then he can refund it and execute program more then block production limit time.

Possibly gas_allowance_counter should not be touched.